src/Markup/Avalonia.Markup.Xaml.Loader/CompilerExtensions/XamlIlBindingPathHelper.cs

```diff
@@ -245,12 +246,18 @@ private static XamlIlBindingPathNode TransformBindingPath(AstTransformationConte
                         nodes.Add(new FindVisualAncestorPathElementNode(visualAncestor.Type, visualAncestor.Level));
                         break;
                     case TemplatedParentBindingExpressionNode templatedParent:
-                        var templatedParentField = context.GetAvaloniaTypes().StyledElement.GetAllFields()
-                            .FirstOrDefault(f => f.IsStatic && f.IsPublic && f.Name == "TemplatedParentProperty");
-                        nodes.Add(new SelfPathElementNode(selfType));
-                        nodes.Add(new XamlIlAvaloniaPropertyPropertyPathElementNode(
-                            templatedParentField,
-                            templatedParent.Type));
+                        var templatedParentType = context
+                            .ParentNodes()
+                            .OfType<AvaloniaXamlIlTargetTypeMetadataNode>()
+                            .Where(x => x.ScopeType == AvaloniaXamlIlTargetTypeMetadataNode.ScopeTypes.ControlTemplate)
+                            .FirstOrDefault()?.TargetType;
+
+                        if (templatedParentType is null)
+                        {
+                            throw new XamlParseException("A binding with a TemplatedParent RelativeSource has to be in a ControlTemplate.", lineInfo);
```

Comment:
We should use as specific exception as possible.
```suggestion
                            throw new XamlTransformException("A binding with a TemplatedParent RelativeSource has to be in a ControlTemplate.", lineInfo);
```